### DISCUSSION TITLE
RSA_check_key() performance on embedded - security implications of removing?

### DISCUSSION THREAD

**Discussion by zevv** [strongswan/strongswan#982] [Q&A]
Hi all,

We have an embedded product running on a pretty restricted CPU, and we find that the `RSA_check_key()` call in `openssl_rsa_private_key.c` is taking a lot of time to validate largish keys (>= 4096 bit). With 10 keys startup time will grow to over a minute, which forced us to increase the `kill kill` timer in starter to allow a successful start at all.

Technically, it's totally possible to remove the `RSA_check_key()` call, startup in a few seconds and have properly working tunnels; this does not allow strongswan to detect corrupt keys at startup time, but as far as I can tell, this should never ever lead to a decrease of security.

Can anyone elaborate on this a bit? Are there any other reasons for calling `RSA_check_key()` apart from early detection of key corruption, and are there any security implications if we would take this test out in our product?

Thanks.

**Comment by tobiasbrunner** [comment]
> With 10 keys

10 private keys? What scenario requires that?

> Are there any other reasons for calling RSA_check_key() apart from early detection of key corruption, and are there any security implications if we would take this test out in our product?

I don't think so (but I'm no cryptographer). I removed the call, too, in the _openssl-3.0_ branch and didn't replace it with `EVP_PKEY_check()` because I didn't consider it necessary when loading trusted private keys (`RSA_check_key` is still in the legacy path if OpenSSL doesn't parse the complete structure itself, it's the only path where we still have the `RSA` structure anyway).

**Reply by zevv** [reply]
> 10 private keys? What scenario requires that?

Good question, but this is in the customers specification; the scenario could be where the device is used as some central hub connecting to a pool of other devices. We are able to support this just fine with smaller key sizes, but things break down because of the long start time when using 4096bit keys.

> I don't think so (but I'm no cryptographer). I removed the call, too, in the _openssl-3.0_ branch and didn't replace it with `EVP_PKEY_check()` because I didn't consider it necessary when loading trusted private keys (`RSA_check_key` is still in the legacy path if OpenSSL doesn't parse the complete structure itself, it's the only path where we still have the `RSA` structure anyway).

Right, this was our hunch as well; I guess two non-crypographers know more then one, eh?

Thanks for the reply, I'll await if anyone comes up telling me this is a totally unsafe and stupid thing to do, otherwise we might go ahead and take it out.

**Reply by tobiasbrunner** [reply]
> the scenario could be where the device is used as some central hub connecting to a pool of other devices

And one key pair isn't enough to uniquely identify the device against the devices in that pool? Or does the spec require that every connection uses a different key pair?

**Comment by zevv** [comment]
There's a ton of different use cases for this device, and we know that some of our users want to run it in a scenario like this. Given that the key loading is the only limitation at this time, and that after loading the keys run just fine, I am not going to argue with them if what they want to do is the right thing for this device or not. If our devices can do what the customer wants, everybody is happy.
